[WIP] Adding resource_attached_disk

chrisst opened the pull request:
This is a resource which will allow joining a arbitrary compute disk
to a compute instance. This will enable dynamic numbers of disks to
be associated by using counts.

Commit:
Adding resource_attached_disk

This is a resource which will allow joining a arbitrary compute disk
to a compute instance. This will enable dynamic numbers of disks to
be associated by using counts.

rileykarson commented on google/resource_attached_disk.go:
```diff
@@ -0,0 +1,286 @@
+		},
+
+		Schema: map[string]*schema.Schema{
+			"attached_disk": {
```
`attached_disk` and `attached_instance` as names seem redundant -  `attached_disk.attached_disk` and `attached_disk.attached_instance` are just noise vs using `attached_disk.instance` and `attached_disk.disk`.

chrisst replied:
I was trying to follow the [naming convention `google_compute_instance` uses](https://www.terraform.io/docs/providers/google/r/compute_instance.html#attached_disk) for consistency.

I agree it's redundant but maybe renaming the whole resource as [Dana pointed out](https://github.com/chrisst/terraform-provider-google/pull/1/files/f26dca38839c4cd5a38547d24444b8b22719e1f0#r212499236) would help a bit. 

In the end I will probably lean for consistency > redundancy.

rileykarson commented on google/resource_attached_disk.go:
```diff
@@ -0,0 +1,286 @@
+
+	attachedDisk := compute.AttachedDisk{}
+	attachedDisk.DeviceName = diskName
+	// TODO strip the v1 part of the path from this instead of the full url
```
Just in case this todo means you intend to do it by hand - check out `GetRelativePath` if you haven't yet! It should do what you want I think.

chrisst replied:
Oops, that was a stale TODO comment. Below I was hard coding the `v1` when constructing the url, and now I'm just using relative.

rileykarson commented on google/resource_attached_disk.go:
```diff
@@ -0,0 +1,286 @@
+		return err
+	}
+
+	attachedInstance, err := getAttachedInstance(d)
```
We can simplify this `Read` method a lot by moving some of this logic to the `ImportState` method; any successful `Create`, or any successful `Import` will have set this information into state already so we don't need the additional fallbacks in `Read`.

My way of thinking of `ImportState` methods is that they should set all the information that was needed to perform a `Create` request (often project, zone/region, name) . This means `Read` methods are very simple since `GET` requests generally take the same parameters.

chrisst replied:
I think I'm not completely clear on the purpose of the `GET` method. I thought it was to populate as many fields as possible on the resource so they could be captured into the tf state. Which is why I was trying to jam all the data into the resource.

So to help me understand: What is the purpose of a Read method? And maybe it's worth breaking it down for what it's purpose is during Create/Delete/Import and other?

rileykarson replied:
Discussed offline - let me know if you're wondering about anything else though.

danawillow commented on google/provider.go:
```diff
@@ -106,6 +106,7 @@ func Provider() terraform.ResourceProvider {
 			GeneratedComposerResourcesMap,
 			GeneratedResourceManagerResourcesMap,
 			map[string]*schema.Resource{
+				"google_attached_disk":                         resourceAttachedDisk(),
```
Let's call it `google_compute_attached_disk` to align with all the other compute engine resources

danawillow commented on google/resource_attached_disk.go:
```diff
@@ -0,0 +1,286 @@
+	return &schema.Resource{
+		Create: resourceAttachedDiskCreate,
+		Read:   resourceAttachedDiskRead,
+		Update: resourceAttachedDiskUpdate, // TODO remove this method? make sure it's not called? etc
```
Yeah you can just remove it and then set all the fields (except output-only ones) to `ForceNew`

rileykarson commented on google/resource_attached_disk.go:
```diff
@@ -0,0 +1,286 @@
+		return nil
+	}
+
+	if err := d.Set("type", ad.Type); err != nil {
```
Since `Type` is a simple type we probably don't need to guard here. We generally check if `d.Set` fails only on nested types.

danawillow commented on google/resource_attached_disk.go:
```diff
@@ -0,0 +1,286 @@
+	instanceName := GetResourceNameFromSelfLink(instance)
+	diskName := GetResourceNameFromSelfLink(d.Get("attached_disk").(string))
+
+	attachedDisk := compute.AttachedDisk{}
```
nit: it's a bit cleaner to do
```
attachedDisk := compute.AttachedDisk{
  DeviceName: [...],
  Source: [...],
}
```
we also tend to use the pointer form here (`&compute.AttachedDisk`), but I don't think there's a real reason for it besides just consistency

rileykarson commented on google/resource_attached_disk.go:
```diff
@@ -0,0 +1,286 @@
+	}
+
+	attachedInstance := d.Get("attached_instance").(string)
+	zone, err := getZoneForAttachedDisk(d, config, attachedInstance)
```
If we have `zone` as a field in state already we can use the one in state - do we have a reason not to just use that one? (This goes for all non `Import`/`Create` uses of `getZoneForAttachedDisk`)

danawillow commented on google/resource_attached_disk_test.go:
```diff
@@ -0,0 +1,100 @@
+// Questions:
+// How do I properly test
+
+func TestAttachedDisk_baisc(t *testing.T) {
```
typo: basic

danawillow commented on google/resource_attached_disk_test.go:
```diff
@@ -0,0 +1,100 @@
+	return nil
+}
+
+func getAll(diskName, instanceName string) string {
```
usually we just name this func the same thing as the test that calls it

danawillow commented on google/resource_compute_disk.go:
```diff
@@ -732,6 +732,7 @@ func resourceComputeDiskUpdate(d *schema.ResourceData, meta interface{}) error {
 }
 
 func resourceComputeDiskDelete(d *schema.ResourceData, meta interface{}) error {
+
```
this change can probably be reverted

danawillow commented on google/resource_attached_disk.go:
```diff
@@ -0,0 +1,286 @@
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%s:%s", instanceName, diskName))
```
If you also put the project and zone in the ID then you can save yourself a bunch of code later in the Read() and Delete() funcs, because then all you need to do to get the info you need would be to parse the ID.

chrisst replied:
🤔 I'm intrigued by this. It would mean only allowing `project/zone/instance:disk` and the more verbose url'y version in the import but I think that's fine.

rileykarson commented on google/resource_attached_disk.go:
```diff
@@ -0,0 +1,286 @@
+	// If zone can't be inferred from the compute instance self link, fall back to project
+	zone, err := getZone(d, c)
+	if err != nil {
+		return "", fmt.Errorf("%s to inherit from the attached instance use `self_url` instead of `name`", err)
```
Is `self_url` meant to be `self_link`?

danawillow commented on google/resource_attached_disk.go:
```diff
@@ -0,0 +1,286 @@
+	config := meta.(*Config)
+
+	// TODO (chrisst) make sure to add good examples to the docs
+	// TODO (chrisst) using 'id' here is a problem. I either need to create a new computed variable (ew)
```
You also don't _have_ to use parseImportId. It might be easier not to.

rileykarson commented on google/self_link_helpers.go:
```diff
@@ -85,3 +85,18 @@ func NameFromSelfLinkStateFunc(v interface{}) string {
+// GetZoneFromSelfLink will attempt to parse the zone if it's in the referenced self link
+//
+// If there is no zone present or the link is malformed it will return an empty string
+func GetZoneFromSelfLink(link string) string {
```
Instead of returning an empty string what if we returned an `error`? That would make it a very explicit part of the function contract that we may fail to return a value.

chrisst replied:
I flip flopped on this one a bunch. I was thinking that the `.Get` methods all just return default values if they haven't been set so was aiming for that behavior. But I suppose since `getZone` returns an error it would be in keeping with that.

rileykarson replied:
Yup!

rileykarson commented on google/resource_attached_disk.go:
```diff
@@ -0,0 +1,286 @@
+	return "", fmt.Errorf("unable to determine the attached compute disk")
+}
+
+func findDiskByID(disks []*compute.AttachedDisk, id string) *compute.AttachedDisk {
```
nit: Won't we always be looking by name here?

chrisst reviewed (commented):
I responded to the comments which I wanted clarification or further discussion on. The rest of them I'll address by doing what was suggested :)

Commit:
Adding code review feedback

rileykarson commented on google/resource_compute_attached_disk.go:
```diff
@@ -0,0 +1,231 @@
+
+	// Force the referenced resources to a self-link in state because it's more specific then name.
+	d.Set("attached_instance", instance.SelfLink)
+	d.Set("attached_disk", fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/zones/%s/disks/%s", project, zone, diskName))
```
You *should* be able to get the self link from instance.Disks? In theory the `Source` attribute is a self link though I  could be very wrong.

chrisst replied:
It's actually an [`AttachedDisk`](https://godoc.org/google.golang.org/api/compute/v1#AttachedDisk) object and not a disk instance. From what I can see it would require a 2nd GET in order to retrieve the actual Disk object.

Commit:
Adding tests and changing timeouts

Added an integration test step to validate that revmoing the attached disk
resource will detach the disk. This is because the concept of 'check destroy'
would only validate that the compute instance was destroyed not that the disk
was detached.

Changed the hard coded timeouts to be configurable

Commit:
Adding acceptance test using count

paddycarver commented on google/resource_compute_attached_disk.go:
```diff
@@ -0,0 +1,227 @@
+		},
+
+		Schema: map[string]*schema.Schema{
+			"attached_disk": {
```
I think `disk` and `instance` make more sense, to me at least.

paddycarver commented on google/resource_compute_attached_disk.go:
```diff
@@ -0,0 +1,227 @@
+				ForceNew:         true,
+				DiffSuppressFunc: compareSelfLinkOrResourceName,
+			},
+			"project": {
```
This should be optional and force new, so the user can attach disks to instances located in projects outside the one they authenticated with. Unless disks and instances are unique across all projects?

paddycarver commented on google/resource_compute_attached_disk.go:
```diff
@@ -0,0 +1,227 @@
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"zone": {
```
Similarly, we probably want the user to be able to specify this, unless disks and instances are both unique across all zones.

chrisst replied:
I left zone & project as computed, because the `attached_disk` resource doesn't really have a project or zone of itself. The zone and project are actually more like properties of the `compute_instance` and are only used to construct the url of the compute instance. 

AFAIK the only reason somebody would need to specify the zone/project in the attached disk resource would be if they were referencing the compute instance by `name` instead of `self_link` (since I parse the self link to compute these) and they hadn't specified a global project+zone. In which case in the error I'm throwing it says "try using self link". 

Does that make sense or am I missing a valid use case for needing to override those variables?

paddycarver replied:
I think that makes perfect sense, thanks for the clarification!

paddycarver commented on google/resource_compute_attached_disk.go:
```diff
@@ -0,0 +1,227 @@
+
+	attachedDisk := compute.AttachedDisk{
+		DeviceName: diskName,
+		Source:     fmt.Sprintf("/projects/%s/zones/%s/disks/%s", project, zone, diskName),
```
Could we use [`getRelativePath`](https://github.com/terraform-providers/terraform-provider-google/blob/10ec7b2ca9a5d191f2eb14b577f2202e2349f6f8/google/self_link_helpers.go#L56) here?

chrisst replied:
If the disk is being referenced by `name` instead of `self_link` then I don't think I have access to the `self_link` here. So I don't think so :(

paddycarver commented on google/resource_compute_attached_disk.go:
```diff
@@ -0,0 +1,227 @@
+	// confirm the disk is actually attached
+	ad := findDiskByName(instance.Disks, diskName)
+	if ad == nil {
+		// Disk was not found attached to the referenced instance
```
We usually like to log a `[WARN]` level log when doing not found checks like this, something to the effect of the disk wasn't found, removing it from state.

paddycarver commented on google/resource_compute_attached_disk.go:
```diff
@@ -0,0 +1,227 @@
+	}
+
+	// Force the referenced resources to a self-link in state because it's more specific then name.
+	d.Set("attached_instance", instance.SelfLink)
```
The [relative link](https://github.com/terraform-providers/terraform-provider-google/blob/10ec7b2ca9a5d191f2eb14b577f2202e2349f6f8/google/self_link_helpers.go#L56) is, I believe, better, as it won't have any diff complications across versions of the API. Though honestly my preference here is weak.

rileykarson replied:
Or self link sets can be wrapped with `ConvertSelfLinkToV1`

chrisst replied:
Looks easy enough, I'm down

paddycarver commented on google/self_link_helpers.go:
```diff
@@ -39,6 +39,7 @@ func compareSelfLinkOrResourceName(_, old, new string, _ *schema.ResourceData) b
 		// `new` is a name
 		// `old` is always a self_link
 		if GetResourceNameFromSelfLink(old) == newParts[0] {
+			// log.Println(" - true!")
```
This should probably be cleaned up.

Commit:
Renaming properties and using relative paths

Commit:
Adding documentation and cleaning comments

Commit:
Zone and Project are now optional

Making the two resources optional so that a user is able to specify a value
without needing to use self links or globally defined values.
Refactored the GetZoneFromLink to work for any path variable (like project)

Commit:
Adding extra configurable properties

Anything that looks like it is passed in as a property of the disk attach
call such as read/write mode, auto delete, device name (mounting), etc.

Commit:
Removing non-hot swappable properties

There are some attach disk properties which can only be used with disks that
are created during the attachDisk call (using initializeParams) so they aren't
useful until we support initialize params.

Commit:
cleaning up syntax

Commit:
Code review feedback

Commit:
Use parseZonalFieldValue instead of custom helpers
